R/test-files.r

```diff
@@ -77,26 +123,42 @@ source_dir <- function(path, pattern = "\\.[rR]$", env = test_env(),
 #' @param path path to file
 #' @param reporter reporter to use
 #' @param env environment in which to execute the tests
-#' @return a data frame of the summary of test results
+#' @return the results as a "testthat_results" (list)
 #' @export
 test_file <- function(path, reporter = "summary", env = test_env()) {
-  current_reporter <- find_reporter(reporter)
+  reporter <- if (is.null(reporter)) NULL else find_reporter(reporter)
+  invisible(.test_file(path, env, reporter, TRUE))
+}
+
+.test_file <- function(path, parent_env, reporter, start_end_reporter) {
+  if (is.null(parent_env)) parent_env <- globalenv()
   lister <- ListReporter$new()
-  reporter <- MultiReporter$new(reporters = list(current_reporter, lister))
-  lister$start_file(basename(path))
-  with_reporter(reporter, .test_file(path, env))
+  reporter <- if (!is.null(reporter))
+    MultiReporter$new(reporters = list(reporter, lister))
+  else
+    lister
 
-  invisible(lister$get_summary())
-}
+  old_reporter <- set_reporter(reporter)
+  old_dir <- setwd(dirname(path))
+  on.exit({
+      setwd(old_dir)
+      set_reporter(old_reporter)
+    })
 
-.test_file <- function(path, parent_env) {
-  old <- setwd(dirname(path))
-  on.exit(setwd(old))
+  if (start_end_reporter) reporter$start_reporter()
 
-  sys.source2(basename(path), new.env(parent = parent_env))
+  fname <- basename(path)
+  lister$start_file(fname)
+
+  sys.source2(fname, new.env(parent = parent_env))
   end_context()
+
+  if (start_end_reporter) reporter$end_reporter()
+
+  invisible(structure(lister$results, class = 'testthat_results'))
```

Comment:
It would be slightly better to have 

```R 
testthat_results  <- function(results) { 
  invisible(results, class = "testthat_results")
}
```

since you can then document that class specification in that function.